File: filter/scripts/fft_export.py
import sys
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from scipy.io import savemat
import matplotlib

matplotlib.use("Agg")  # Headless mode


def run_pipeline(input_csv, filtered_csv):
    # 1. Load & Process
    df_in = pd.read_csv(input_csv)
    df_out = pd.read_csv(filtered_csv)

    t_in, sig_in = df_in["Time"].values, df_in["Amplitude"].values
    t_out, sig_out = df_out["Time"].values, df_out["Amplitude"].values

    def get_fft(sig):
        n = len(sig)
        f = np.fft.rfftfreq(n)
        m = 20 * np.log10(np.abs(np.fft.rfft(sig)) + 1e-9)
        return f, m

    f_in, m_in = get_fft(sig_in)
    f_out, m_out = get_fft(sig_out)

    # 2. Save Data to .mat (The Bridge)
    mat_name = filtered_csv.replace(".csv", ".mat")
    data_dict = {
        "t_in": t_in,
        "sig_in": sig_in,
        "f_in": f_in,
        "m_in": m_in,
        "t_out": t_out,
        "sig_out": sig_out,
        "f_out": f_out,
        "m_out": m_out,
        "source": filtered_csv,
    }
    savemat(mat_name, data_dict)

    # 3. Save Static PNG
    plt.figure(figsize=(10, 8))
    plt.subplot(2, 1, 1)
    plt.plot(t_in, sig_in, alpha=0.3)
    plt.plot(t_out, sig_out)
    plt.subplot(2, 1, 2)
    plt.plot(f_in, m_in, alpha=0.3)
    plt.plot(f_out, m_out)
    plt.savefig(filtered_csv.replace(".csv", ".png"))
    plt.close()
    print(f"Processed {filtered_csv} -> .mat and .png")


if __name__ == "__main__":
    run_pipeline(sys.argv[1], sys.argv[2])

File: filter/scripts/fft_export.m
arg_list = argv();
in_file = arg_list{1}; filt_file = arg_list{2};

% Load and Compute
d_in = csvread(in_file, 1, 0); t_in = d_in(:,1); sig_in = d_in(:,2);
d_out = csvread(filt_file, 1, 0); t_out = d_out(:,1); sig_out = d_out(:,2);

function [f, m] = do_fft(s)
    N = length(s); f = linspace(0, 0.5, floor(N/2)+1);
    m = 20*log10(abs(fft(s))(1:length(f)) + 1e-9);
endfunction

[f_in, m_in] = do_fft(sig_in); [f_out, m_out] = do_fft(sig_out);

% Save Bridge File
mat_name = strrep(filt_file, '.csv', '.mat');
save("-mat7-binary", mat_name, "t_in", "sig_in", "f_in", "m_in", "t_out", "sig_out", "f_out", "m_out");

% Save PNG
h = figure('visible', 'off');
subplot(2,1,1); plot(t_in, sig_in, 'color', [0.8 0.8 0.8], t_out, sig_out, 'b');
subplot(2,1,2); plot(f_in, m_in, 'color', [0.8 0.8 0.8], f_out, m_out, 'r');
print(h, strrep(filt_file, '.csv', '.png'), '-dpng');
